src/Autofac.Diagnostics.DotGraph/DotDiagnosticTracer.cs

```diff
@@ -0,0 +1,558 @@
+        {
+            if (data is null)
+            {
+                return;
+            }
+
+            if (_operationBuilders.TryGetValue(data.Operation, out var builder))
+            {
+                try
+                {
+                    builder.OnOperationFailure();
+                    OnOperationCompleted(new OperationTraceCompletedArgs<string>(data.Operation, builder.ToString()));
+                }
+                finally
+                {
+                    _operationBuilders.TryRemove(data.Operation, out var _);
+                }
+            }
+        }
+
+        /// <inheritdoc/>
+        protected override void OnRequestStart(RequestDiagnosticData data)
+        {
+            if (data is null)
+            {
+                return;
+            }
+
+            if (_operationBuilders.TryGetValue(data.Operation, out var builder))
+            {
+                builder.OnRequestStart(
+                    data.RequestContext.Service,
+                    data.RequestContext.Registration.Activator.DisplayName(),
+                    data.RequestContext.DecoratorTarget?.Activator.DisplayName());
+            }
+        }
+
+        /// <inheritdoc/>
+        protected override void OnRequestSuccess(RequestDiagnosticData data)
+        {
+            if (data is null)
+            {
+                return;
+            }
+
+            if (_operationBuilders.TryGetValue(data.Operation, out var builder))
+            {
+                builder.OnRequestSuccess(data.RequestContext.Instance);
+            }
+        }
+
+        /// <inheritdoc/>
+        protected override void OnRequestFailure(RequestFailureDiagnosticData data)
+        {
+            if (data is null)
+            {
+                return;
+            }
+
+            if (_operationBuilders.TryGetValue(data.Operation, out var builder))
+            {
+                var requestException = data.RequestException;
+                if (requestException is DependencyResolutionException && requestException.InnerException is object)
+                {
+                    requestException = requestException.InnerException;
+                }
+
+                if (requestException.Data.Contains(RequestExceptionTraced))
+                {
+                    builder.OnRequestFailure(null);
+                }
+                else
+                {
+                    builder.OnRequestFailure(requestException);
+                }
+
+                requestException.Data[RequestExceptionTraced] = true;
+            }
+        }
+
+        /// <summary>
+        /// One node in the graph. The resulting instance (for successful requests)
+        /// is what uniquely identifies the node when normalizing the data. This
+        /// converts the notion of a "resolve request" into a dependency graph
+        /// based on completed resolutions.
+        /// </summary>
+        private class ResolveRequestNode
```

Comment:
There's quite a lot of private classes in here; might these be better as internal classes in their own files, for readability purposes?

Reply:
Oh, good point. Let me fix that.

Reply:
Pushed.